### PR TITLE
chore: wrap training library with try/catch handles

### PR DESCRIPTION
**Pull request by RobotSail**
Previously, the instructlab/training run_training function would handle errors
by silently failing and reporting a success. As of training PR #217, we have fixed
this behaior by now raising a RuntimeError when we encounter a problem inside of
the training loop. This commit handles this error inside of the accelerated and
phased training pipelines.

Signed-off-by: Oleg S <97077423+RobotSail@users.noreply.github.com>

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/model/train.py
Comment (danmcp) [inline src/instructlab/model/train.py:651]: I think `raise click.exceptions.Exit(1)` is preferred

File: src/instructlab/model/train.py
Comment (danmcp) [inline src/instructlab/model/train.py:1092]: Seems like this new logic wrapping _training_phase could be shared.  Either in _training_phase or in a new function.

**Review by danmcp — CHANGES_REQUESTED** [review]

File: src/instructlab/model/train.py
Comment (RobotSail) [reply src/instructlab/model/train.py:651]: Was not aware of this API, thank you!

File: src/instructlab/model/train.py
Comment (RobotSail) [reply src/instructlab/model/train.py:1092]: Is this necessary? It seems like the duplication is only happening twice, and so the other section may need to change or do something different in the future. I'm concerned about premature abstraction.

Comment (danmcp) [reply src/instructlab/model/train.py:1092]: My vote would be to abstract it now because:
- With the similar code duplicated, it's really easy for one section to migrate from the other over time.  Wouldn't want that to happen accidentially.
- It's easier to figure out what's duplicated now vs as the code evolves over time

Comment (RobotSail) [reply src/instructlab/model/train.py:1092]: Thank you for that explanation, that makes sense. I'll make the change.

Comment (RobotSail) [reply src/instructlab/model/train.py]: I've updated this, let me know if it aligns with your suggestion or if it needs to be updated.

**Review by cdoern — COMMENTED** [review]
can we hold this until #2436 merges? this will cause massive issues with the client-server work

File: src/instructlab/model/train.py
Comment (danmcp) [inline src/instructlab/model/train.py:1089]: I think this should be a click.echo or a print.  The line being, if logging was disabled or sent to a log, would you still want this message to be printed.

File: src/instructlab/model/train.py
Comment (danmcp) [reply src/instructlab/model/train.py]: Looks great, thanks!

File: src/instructlab/model/train.py
Comment (RobotSail) [reply src/instructlab/model/train.py:1089]: The logger is used here since I saw it being used elsewhere in this script. I'll update it to use click.secho though for broader consistency.

**Comment by RobotSail** [comment]
Blocked by #2513

**Comment by RobotSail** [comment]
@mergifyio rebase

**Review by danmcp — APPROVED** [review]

File: src/instructlab/model/accelerated_train.py
Comment (JamesKunstle) [inline src/instructlab/model/accelerated_train.py:286]: Could you please use arg names rather than positional?
e.g.
```suggestion
        _run_phase(train_args=train_args, torch_args=torch_args, data_path=phase1_data, checkpoint_dir=phase1_checkpoints_dir, ...)
```

File: src/instructlab/model/accelerated_train.py
Comment (JamesKunstle) [inline src/instructlab/model/accelerated_train.py:290]: this should be the checkpoints dir

File: src/instructlab/model/accelerated_train.py
Comment (JamesKunstle) [inline src/instructlab/model/accelerated_train.py:235]: I think this indentation might be wrong? or this might be a try/catch syntax thing I only barely remember

File: src/instructlab/model/accelerated_train.py
Comment (JamesKunstle) [inline src/instructlab/model/accelerated_train.py:493]: instead of _run_phase including bookkeeping, we could remove the bookkeeping and reuse that function here.

**Review by JamesKunstle — CHANGES_REQUESTED** [review]
could of design requests

File: src/instructlab/model/accelerated_train.py
Comment (RobotSail) [reply src/instructlab/model/accelerated_train.py:235]: This is a try/catch specific syntax. The `else` here is basically saying "if I've caught an exception, handle it in these except blocks, else run this code"

File: src/instructlab/model/accelerated_train.py
Comment (RobotSail) [reply src/instructlab/model/accelerated_train.py:238]: @danmcp Recommended extracting this pattern out.

Comment (RobotSail) [reply src/instructlab/model/accelerated_train.py:238]: @JamesKunstle Please see this conversation: https://github.com/instructlab/instructlab/pull/2476#discussion_r1805093423

File: src/instructlab/model/accelerated_train.py
Comment (RobotSail) [reply src/instructlab/model/accelerated_train.py:290]: Good catch!

File: src/instructlab/model/accelerated_train.py
Comment (RobotSail) [reply src/instructlab/model/accelerated_train.py:493]: I see your point here but I'd rather not do this. The point of `_run_phase` was to simplify the flow within the multi-phase training. @danmcp maybe you had some thoughts here?

Comment (danmcp) [reply src/instructlab/model/accelerated_train.py:493]: I am probably missing something, but looking at the current version closer, I am not sure why we need the try/except logic here at all.  Isn't this only called by _run_phase which also has the RuntimeError handling added?  There is a separate call to `run_training` in `accelerated_train` which does seem like it needs the exception handling instead.

**Review by danmcp — CHANGES_REQUESTED** [review]

File: src/instructlab/model/accelerated_train.py
Comment (RobotSail) [reply src/instructlab/model/accelerated_train.py:493]: @danmcp Good point, I'll remove this line altogether.

Comment (RobotSail) [reply src/instructlab/model/accelerated_train.py]: @JamesKunstle @danmcp I've removed this try/catch and moved it to the other `run_training` invocation per your suggestions. Please take a look and let me know if you spot any problems.

File: src/instructlab/model/accelerated_train.py
Comment (JamesKunstle) [reply src/instructlab/model/accelerated_train.py:240]: ++

File: src/instructlab/model/accelerated_train.py
Comment (JamesKunstle) [reply src/instructlab/model/accelerated_train.py:243]: Yeah I see the point very clearly, having duplicated code for updating the journal is unsightly. Maybe two functions would be more reasonable:

```
def _run_phase(journal, next_phase, phase_func):
    outputs = phase_func()
    journal.current_phase = next_phase
    journal.commit()
    return outputs

def _handle_training_failures(train_func):
    try:
        train_func()
     else:
        ...
    except: 
        ...
```

**Review by danmcp — APPROVED** [review]

File: src/instructlab/model/accelerated_train.py
Comment (JamesKunstle) [reply src/instructlab/model/accelerated_train.py:243]: I can see that this isn't a super nice way to do things on inspection. We can potentially refactor this later but it'll work for the time being.

**Review by JamesKunstle — APPROVED** [review]

**Pull request merged**
